crates/red_knot_python_semantic/src/types/diagnostic.rs

```diff
@@ -987,10 +987,11 @@ pub(super) fn report_slice_step_size_zero(context: &InferContext, node: AnyNodeR
 pub(super) fn report_invalid_assignment(
     context: &InferContext,
     node: AnyNodeRef,
-    declared_ty: Type,
-    assigned_ty: Type,
+    target_ty: Type,
+    source_ty: Type,
+    target_description: Option<&'static str>,
```

Comment:
Let me know if you prefer another function for this. I the `target_description` parameter in an attempt to get better error messages in even more cases, but haven't really found any so far.